### ISSUE TITLE
Few key actions not working when wc3270 is launched using script

### ISSUE THREAD

**Issue by FWDQK** [Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library#69]
Hi Team,
We are working on RobotFramework using mainframe library.
When script runs 'Open Connection' keyword and launches wc3270, it does not allow to hit F3 key.
We have tried manually also on the terminal  window launched by script, and it gives an error 'Function key is not allowed'

F3 key is used to exit from current page.

Same issue observed for page down and other selected keys.

**Comment by robinmatz** [comment]
Hi @FWDQK ,

thank you for bringing this to our attention.

Could you please paste a copy of the script or provide a minimal example so we can try to reproduce the behavior you described?

**Comment by FWDQK** [comment]
[LARF code.txt](https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/files/9442659/LARF.code.txt)
HI,
This is the sample code we are trying to launch AS400 and login.
Post login, when we are trying to hit F3, it gives an error Function Key not allowed.

**Comment by FWDQK** [comment]
Hi,
I have also tried to perform below keyword, but still it gives an error :
Execute Command     PF(3)

**Comment by samuelpcabral** [comment]
Hello @FWDQK , unfortunately at the moment I only have access to the mainframe at pub400.com and there "Send PF" is working normally. Is it not related to some action "stuck" inside the mainframe you are using?

**Comment by FWDQK** [comment]
> Hello @FWDQK , unfortunately at the moment I only have access to the mainframe at pub400.com and there "Send PF" is working normally. Is it not related to some action "stuck" inside the mainframe you are using?



> Hello @FWDQK , unfortunately at the moment I only have access to the mainframe at pub400.com and there "Send PF" is working normally. Is it not related to some action "stuck" inside the mainframe you are using?

It is issue observed inside the mainframe used via emulator.
Via RobotFramework, it launches wc3270 exe. Post, Open connection and login via script, it we are trying to perform few actions manually also, it is not working.

But when we are trying to launch wc3270 exe manually and perform flow, its working fine.

It is similar to below issue :
https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/issues/20

**Comment by FWDQK** [comment]
Just wanted to check if we can pass both pub400.com and hostname (ip address) to open connection?

Can you please help to share code used to launch pub400.com.

We are using as400 IBM mainframe application.

**Comment by FWDQK** [comment]
Hi,
Please help to confirm how we can do keyboard configuration on wc3270.exe.

**Comment by samuelpcabral** [comment]
Sorry, but what do you mean by keyboard configuration? Our library just sends the commands to the wc3270 terminal so that Robot can perform the basic actions.

Anyway, did you take a look at this link? https://x3270.miraheze.org/wiki/Wc3270

**Comment by samuelpcabral** [comment]
sorry, never used those settings... can't help.

**Comment by robinmatz** [comment]
Hey @FWDQK ,

I have followed this conversation but haven't had the time to respond yet.

In my point of view, we have to figure out if the problem you are facing on the x3270's side or on Mainframe3270 library.


From what I've read in this thread you are saying that the behavior does not occur when running wc3270 manually and connecting from there, correct?



If so, I would like to ask you to one by one launch wc3270 with the options that are provided to wc3270 by Mainframe3270 library.

Those would be 
`        self.args = ["-xrm", "wc3270.unlockDelay: False", "-xrm", "wc3270.model: 2"]`


So, basically if you launch like `wc3270 -xrm wc3270.model: 2` does the error occur, too?

**Comment by FWDQK** [comment]
> atest

Hi,
Yes, correct. It is working fine if we are trying to execute wc3270.exe manually.
But facing issue when tried to launch via Robot framework script.

I have tried to add below code, but getting an error : WinError 10061] No connection could be made because the target machine actively refused it

${listarg}      Create List     "-xrm"     "wc3270.unlockDelay: False"   "-xrm"  "wc3270.model: 2"
Open Connection    ${mainframe_host}        extra_args=${listarg}

Please suggest if code syntax is correct or not.

**Comment by samuelpcabral** [comment]
Hi, in Robot Framework you don´t put the quotes for the strings, so for the create list is something like this:
Use the log keyword to see if the list is created in python as should be **["-xrm", "wc3270.unlockDelay: False", "-xrm", "wc3270.model: 2"]**

```
${listarg}     Create List    -xrm    wc3270.unlockDelay: False    -xrm    wc3270.model: 2
Log    ${listarg}    WARN
```

**Comment by FWDQK** [comment]
> ${listarg}     Create List    -xrm    wc3270.unlockDelay: False    -xrm    wc3270.model: 2
> Log    ${listarg}    WARN

Thanks.
Tried this, but still the same issue observed.
Unable to use F3 and few keys.
wc3270 still gives an error Function Key not allowed.

**Comment by robinmatz** [comment]
Hi @FWDQK ,

I think you got the wrong idea.
I was not talking about adding the above options in the robot script.

What I meant was for you to open wc3270 from command line prompt with the additional options.

**Comment by FWDQK** [comment]
> wc3270 -xrm wc3270.model: 2

Can you please help with the exact command, because while trying, it is giving an error.

**Comment by robinmatz** [comment]
> > wc3270 -xrm wc3270.model: 2
> 
> Can you please help with the exact command, because while trying, it is giving an error.

Almost, I think you need to remove the blanks pace between ':' and '2'.

`wc3270 -xrm wc3270.model:2`

**Comment by FWDQK** [comment]
> wc3270 -xrm wc3270.model:2

If this is used to launch wc3270.exe manually, it throws an error Function Key not allowed.
Same error which is faced via script.

Please suggest how we can resolve this via script also.

**Comment by robinmatz** [comment]
Ok, so now we know what is causing the issue, I need some time to think about a solution

**Comment by robinmatz** [comment]
Hi @FWDQK ,

as for now as a workaround, I suggest you do the following

```robot
*** Settings ***
Library    Mainframe3270


*** Test Cases ***
Connect
    ${extra_args} =    Create List    -xrm    wc3270.model:4
     Open Connection    your_host.com    extra_args=${extra_args}
```

As you can see, in the [x3270 wiki](https://x3270.miraheze.org/wiki/Model_resource) the default model for wc3270 is `3279-4-E`, which can be abbreviated to just `4`. By passing in the `${extra_args}` from the abbove code snippet, we override Robot-Framework-Mainframe3270-Library's behavior to use model 2. In other words, we choose the default model also in robot scripts.

Please let me know if this approache solves the issue for you.

**Comment by FWDQK** [comment]
> wc3270 -xrm wc3270.model:2

Hi,
Manually if tried to launch using wc3270 -xrm wc3270.model:4, its working.
But if we have tried same using script, it types data in first field were cursor has default focus. 
Write in position       ${mainframe_password}      7   53
It is not writing as per the position mentioned.
Even manually in that window, I am unable to perform any action to login and verify it manually if launched via script, we can perform required flow.

**Comment by robinmatz** [comment]
Hi @FWDQK , I raised an issue in the x3270 repository.

**Comment by robinmatz** [comment]
Hi @FWDQK ,

so this is the answer from x3270's maintainer, Paul Mattes:


>Here is the big difference: With a model 2, the host accepts that the terminal is a real 3270 and the session uses the 3270 data stream protocol (what is what wc3270 is intended for). With a model 4, the host is apparently unhappy with the protocol negotiation and leaves the session in TELNET NVT mode, where wc3270 is emulating an xterm (ANSI/VT100-like serial terminal).


>What the AS/400 host really wants to talk to is an IBM 5250 terminal. The host has a 5250 translation layer for the TELNET session, and it looks like the PF key mappings are different between the 3270-to-5250 translations (used for model 2) and the VT100-to-5250 translations (used for model 4).


>For a 3270, the following document applies: https://www.ibm.com/docs/en/i/7.2?topic=sessions-3270-keyboard-mapping-telnet-servers

>As you can see in that document, the 3270 PF4 key is actually the 5250 'Print' function, which the trace shows is how it was interpreted. If you want the 5250 F4 key, you need to press PA1, followed by PF4, on wc3270.


>For a VT100 terminal, the mappings are direct: PF4 is interpreted as the 5250 F4 key. Of course, that means that the 5250 Print function is not available.


So what does that mean for your use case? Given the above explanation applies to your host, in order to Send PF 1-12, you will have to do send PA(1) before sending the PF keys:

```robot
*** Settings ***
Library     Mainframe3270


*** Test Cases ***
Send PF
    Open Connection     yourhost.com
    Execute Command     PA(1)
    Send PF    3
```

Other mappings can be found in the document referenced by Paul: https://www.ibm.com/docs/en/i/7.2?topic=sessions-3270-keyboard-mapping-telnet-servers


Please let me know if this resolves the issue for you.

**Comment by FWDQK** [comment]
> Hi @FWDQK ,
> 
> so this is the answer from x3270's maintainer, Paul Mattes ([pmattes/x3270#48 (comment)](https://github.com/pmattes/x3270/issues/48#issuecomment-1249529975)):
> 
> > Here is the big difference: With a model 2, the host accepts that the terminal is a real 3270 and the session uses the 3270 data stream protocol (what is what wc3270 is intended for). With a model 4, the host is apparently unhappy with the protocol negotiation and leaves the session in TELNET NVT mode, where wc3270 is emulating an xterm (ANSI/VT100-like serial terminal).
> 
> > What the AS/400 host really wants to talk to is an IBM 5250 terminal. The host has a 5250 translation layer for the TELNET session, and it looks like the PF key mappings are different between the 3270-to-5250 translations (used for model 2) and the VT100-to-5250 translations (used for model 4).
> 
> > For a 3270, the following document applies: https://www.ibm.com/docs/en/i/7.2?topic=sessions-3270-keyboard-mapping-telnet-servers
> 
> > As you can see in that document, the 3270 PF4 key is actually the 5250 'Print' function, which the trace shows is how it was interpreted. If you want the 5250 F4 key, you need to press PA1, followed by PF4, on wc3270.
> 
> > For a VT100 terminal, the mappings are direct: PF4 is interpreted as the 5250 F4 key. Of course, that means that the 5250 Print function is not available.
> 
> So what does that mean for your use case? Given the above explanation applies to your host, in order to Send PF 1-12, you will have to do send PA(1) before sending the PF keys:
> 
> ```
> *** Settings ***
> Library     Mainframe3270
> 
> 
> *** Test Cases ***
> Send PF
>     Open Connection     yourhost.com
>     Execute Command     PA(1)
>     Send PF    3
> ```
> 
> Other mappings can be found in the document referenced by Paul: https://www.ibm.com/docs/en/i/7.2?topic=sessions-3270-keyboard-mapping-telnet-servers
> 
> Please let me know if this resolves the issue for you.

Hi @robinmatz ,
Thanks a lot for your help.
It worked by Execute Command PA(1), before F3.

I have few questions, if you can help with :
1. I want to click on particular row and col, is there any direct keyword for setcursor position?
2. Current screenshot keyword, captures screenshot in html format. Do we have any option for jpg or png format? Also provide customized screenshot file name

**Comment by robinmatz** [comment]
Hi @FWDQK 

sorry for the late reply.

1: There is no direct keyword for this yet. You can however use this,
`Execute Command    MoveCursor(y, x)`
where y is the position of the y-axis and x the position of the x-axis. Those values are 0-based, so MoveCursor(5, 15) will move the cursor to 6, 16.

If you would like to have it in the Mainframe3270 library, you can open an issue and we will implement it.


2: Unfortunatly, we do not support png or jpg. This is because the PrintText() from x3270 that we use internally, also does not support jpn or png (https://x3270.miraheze.org/wiki/PrintText()_action)

3: Keys like page up or page down usually have a corresponding PF key, e.g. PF(7) for page up or PF(8) for page down. You can use these instead.

**Comment by FWDQK** [comment]
> Hi @FWDQK
> 
> sorry for the late reply.
> 
> 1: There is no direct keyword for this yet. You can however use this, `Execute Command MoveCursor(y, x)` where y is the position of the y-axis and x the position of the x-axis. Those values are 0-based, so MoveCursor(5, 15) will move the cursor to 6, 16.
> 
> If you would like to have it in the Mainframe3270 library, you can open an issue and we will implement it.
> 
> 2: Unfortunatly, we do not support png or jpg. This is because the PrintText() from x3270 that we use internally, also does not support jpn or png (https://x3270.miraheze.org/wiki/PrintText()_action)
> 
> 3: Keys like page up or page down usually have a corresponding PF key, e.g. PF(7) for page up or PF(8) for page down. You can use these instead.

Hi,
Thanks for revert.
For Point 1, we will try and confirm.
For point 2, wanted to check if we can maximize wc3270.
For point 3, we have already tried PF7 and PF8, but it is not working.
It was showing message below X scrolled 1 and so on. But it does not perform action. Tried also to apply Execute Command PA(1) too.

**Comment by FWDQK** [comment]
> > Hi @FWDQK
> > sorry for the late reply.
> > 1: There is no direct keyword for this yet. You can however use this, `Execute Command MoveCursor(y, x)` where y is the position of the y-axis and x the position of the x-axis. Those values are 0-based, so MoveCursor(5, 15) will move the cursor to 6, 16.
> > If you would like to have it in the Mainframe3270 library, you can open an issue and we will implement it.
> > 2: Unfortunatly, we do not support png or jpg. This is because the PrintText() from x3270 that we use internally, also does not support jpn or png (https://x3270.miraheze.org/wiki/PrintText()_action)
> > 3: Keys like page up or page down usually have a corresponding PF key, e.g. PF(7) for page up or PF(8) for page down. You can use these instead.
> 
> Hi, Thanks for revert. For Point 1, we will try and confirm. For point 2, wanted to check if we can maximize wc3270. For point 3, we have already tried PF7 and PF8, but it is not working. It was showing message below X scrolled 1 and so on. But it does not perform action. Tried also to apply Execute
[wc3270 trace for page down.zip](https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/files/9642843/wc3270.trace.for.page.down.zip)
 Command PA(1) too.
[wc3270 trace for page down.zip](https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/files/9642854/wc3270.trace.for.page.down.zip)

Hi,

Point 1 is working fine for MoveCursor.
Waiting revert for Point 2, if we can maximize wc3270 window using any key or action.
Point 3, I have attached screenshot and trace report for reference where tried to hit via PF7 with and without Execute Command.

**Comment by FWDQK** [comment]
> > > Hi @FWDQK
> > > sorry for the late reply.
> > > 1: There is no direct keyword for this yet. You can however use this, `Execute Command MoveCursor(y, x)` where y is the position of the y-axis and x the position of the x-axis. Those values are 0-based, so MoveCursor(5, 15) will move the cursor to 6, 16.
> > > If you would like to have it in the Mainframe3270 library, you can open an issue and we will implement it.
> > > 2: Unfortunatly, we do not support png or jpg. This is because the PrintText() from x3270 that we use internally, also does not support jpn or png (https://x3270.miraheze.org/wiki/PrintText()_action)
> > > 3: Keys like page up or page down usually have a corresponding PF key, e.g. PF(7) for page up or PF(8) for page down. You can use these instead.
> > 
> > 
> > Hi, Thanks for revert. For Point 1, we will try and confirm. For point 2, wanted to check if we can maximize wc3270. For point 3, we have already tried PF7 and PF8, but it is not working. It was showing message below X scrolled 1 and so on. But it does not perform action. Tried also to apply Execute
> > [wc3270 trace for page down.zip](https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/files/9642843/wc3270.trace.for.page.down.zip)
> > Command PA(1) too.
> > [wc3270 trace for page down.zip](https://github.com/Altran-PT-GDC/Robot-Framework-Mainframe-3270-Library/files/9642854/wc3270.trace.for.page.down.zip)
> 
> Hi,
> 
> Point 1 is working fine for MoveCursor. Waiting revert for Point 2, if we can maximize wc3270 window using any key or action. Point 3, I have attached screenshot and trace report for reference where tried to hit via PF7 with and without Execute Command.

Adding Point 4, if any invalid key gets hit, it shows as Protected.
How can we come out of it. In as400, we used to click control to reset and make screen active.
![image](https://user-images.githubusercontent.com/111964982/192195352-6c75cbe8-6d46-4aa9-8c1f-229b32b489c5.png)

**Comment by FWDQK** [comment]
> if any invalid key gets hit, it shows as Protected.
> How can we come out of it. In as400, we used to click control to reset and make screen active.

Hi,
Able to access page down and page up now.

**Comment by robinmatz** [comment]
> Hi,
   Able to access page down and page up now.

Hey @FWDQK ,

great to hear that! Can you tell us how you resolved the issue?

**Comment by FWDQK** [comment]
> > Hi,
> > Able to access page down and page up now.
> 
> Hey @FWDQK ,
> 
> great to hear that! Can you tell us how you resolved the issue?

PF 8 was for page down and PF 7 for page up.

How to handle X protected window?

**Comment by FWDQK** [comment]
> > > Hi,
> > > Able to access page down and page up now.
> > 
> > 
> > Hey @FWDQK ,
> > great to hear that! Can you tell us how you resolved the issue?
> 
> PF 8 was for page down and PF 7 for page up.
> 
> How to handle X protected window?

How to pass alt + r via script?

**Comment by samuelpcabral** [comment]
@FWDQK Hi, solved this? I think you can use any python package that handle with keyboard actions.

**Comment by robinmatz** [comment]
Hi @FWDQK 

having seen you also posted this on x3270, when trying to execute a x3270 action that has no dedicated keyword in this library, you can use `Execute Command`. In your case you were trying to execute the `Reset()` action. In Robot Framework this will look like

```robot
Execute Command    Reset()
```

**Comment by robinmatz** [comment]
Since the original question has been answered and there has not been any recent activity, I am going to close this issue.
